cpp/src/arrow/util/concatenate.cc

```diff
@@ -0,0 +1,275 @@
+  }
+
+  Status ConcatenateBitmaps(const std::vector<Bitmap>& bitmaps,
+                            std::shared_ptr<Buffer>* out) {
+    RETURN_NOT_OK(AllocateBitmap(pool_, out_.length, out));
+    uint8_t* bitmap_data = (*out)->mutable_data();
+    int64_t bitmap_offset = 0;
+    for (int i = 0; i < in_size(); ++i) {
+      auto bitmap = bitmaps[i];
+      if (bitmap.AllSet()) {
+        BitUtil::SetBitsTo(bitmap_data, bitmap_offset, bitmap.range.length, true);
+      } else {
+        internal::CopyBitmap(bitmap.data, bitmap.range.offset, bitmap.range.length,
+                             bitmap_data, bitmap_offset, false);
+      }
+      bitmap_offset += bitmap.range.length;
+    }
+    if (auto preceding_bits = BitUtil::kPrecedingBitmask[out_.length % 8]) {
+      bitmap_data[out_.length / 8] &= preceding_bits;
+    }
+    return Status::OK();
+  }
+
+  Status PutOffsets(const std::shared_ptr<Buffer>& src, Range range,
+                    int32_t values_length, int32_t* dst_begin, Range* values_range) {
+    auto src_begin = reinterpret_cast<const int32_t*>(src->data()) + range.offset;
+    auto src_end = src_begin + range.length;
+    auto first_offset = src_begin[0];
+    values_range->offset = first_offset;
+    values_range->length = *src_end - values_range->offset;
+    if (values_length > std::numeric_limits<int32_t>::max() - values_range->length) {
+      return Status::Invalid("offset overflow while concatenating arrays");
+    }
+    std::transform(src_begin, src_end, dst_begin,
+                   [values_length, first_offset](int32_t offset) {
+                     return offset - first_offset + values_length;
+                   });
+    return Status::OK();
+  }
+
+  Status ConcatenateOffsets(const BufferVector& buffers, std::shared_ptr<Buffer>* out,
+                            std::vector<Range>* ranges) {
+    RETURN_NOT_OK(AllocateBuffer(pool_, (out_.length + 1) * sizeof(int32_t), out));
+    auto dst = reinterpret_cast<int32_t*>((*out)->mutable_data());
+    int64_t elements_length = 0;
+    int32_t values_length = 0;
+    ranges->resize(in_size());
+    for (int i = 0; i < in_size(); ++i) {
+      RETURN_NOT_OK(PutOffsets(buffers[i], in_range(i), values_length,
+                               &dst[elements_length], &ranges->at(i)));
+      elements_length += in_length(i);
+      values_length += static_cast<int32_t>(ranges->at(i).length);
+    }
+    dst[elements_length] = values_length;
+    return Status::OK();
+  }
+
+  int in_size() const { return static_cast<int>(in_.size()); }
+  int64_t in_offset(int i) const { return in_[i].offset; }
+  int64_t in_length(int i) const { return in_[i].length; }
+  Range in_range(int i) const { return Range(in_offset(i), in_length(i)); }
+
+  const std::vector<ArrayData>& in_;
+  MemoryPool* pool_;
+  ArrayData out_;
+};
+
+Status Concatenate(const ArrayVector& arrays, MemoryPool* pool,
+                   std::shared_ptr<Array>* out) {
+  DCHECK_GT(arrays.size(), 0);
```

Comment:
Okay, I'll return `Invalid`. Should I add that overload too?

Reply:
I think it would be nice (and easy) to add that overload. It's a bit annoying to have APIs that disallow 0-sized inputs, because the caller always has to ensure upfront that they're respecting the contract.